patches/server/0907-Add-ItemStack-damage.patch

```diff
@@ -0,0 +1,38 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Sun, 8 May 2022 13:35:45 -0700
+Subject: [PATCH] Add ItemStack#damage
+
+Adds methods to simulate damage done to an itemstack
+and all the logic associated with damaging them
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+index c022751e3b45469cc0ad6732e2d6ff08918bafa4..6ccd36537d866113ddb5734d31618bb47dcb763f 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+@@ -919,5 +919,25 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
+         }
+         throw new IllegalArgumentException(entityCategory + " is an unrecognized entity category");
+     }
++
++    @Override
++    public ItemStack damageItemstack(ItemStack stack, int amount, org.bukkit.inventory.EquipmentSlot slot) {
++        final net.minecraft.world.item.ItemStack nmsStack;
++        if (stack instanceof CraftItemStack craftItemStack) {
++            if (craftItemStack.handle == null || craftItemStack.handle.isEmpty()) {
++                return stack;
++            }
++            nmsStack = craftItemStack.handle;
++        } else {
++            nmsStack = CraftItemStack.asNMSCopy(stack);
++            stack = CraftItemStack.asCraftMirror(nmsStack); // mirror to capture changes in hurt logic & events
++        }
++        nmsStack.hurtAndBreak(amount, this.getHandle(), livingEntity -> {
++            if (slot != null) {
++                livingEntity.broadcastBreakEvent(org.bukkit.craftbukkit.CraftEquipmentSlot.getNMS(slot));
```

Comment:
what happens if I just pass the wrong slot?

Reply:
I'm not sure what you mean the "wrong" slot?

Reply:
probably by doing something like damageItemstack(player.getInventory().getItemInMainHand(), 1, EquipmentSlot.FEET);

Reply:
Yeah, that just tells all the clients to play the break animation and sound for the feet slot. it doesn't change any client state. so it doesn't really matter.

Reply:
Loop the players armor and hand items to check for the matching slot for the ItemStack given 5Head - Would a `damageItemstack(LivingEntity, Int, EquipmentSlot)` make sense to damage any ItemStack in that slot?

Reply:
Oh that does look better. instead of having the nullable slot, just have a method for just a stack, and one for a slot